README.md
![license](https://img.shields.io/github/license/hhk7734/tensorflow-yolov4)
![pypi](https://img.shields.io/pypi/v/yolov4)
![language](https://img.shields.io/github/languages/top/hhk7734/tensorflow-yolov4)

# tensorflow-yolov4

YOLOv4 Implemented in Tensorflow 2.

Docs: [https://wiki.loliot.net/docs/lang/python/libraries/yolov4/python-yolov4-about](https://wiki.loliot.net/docs/lang/python/libraries/yolov4/python-yolov4-about)

## Installation

```shell
python3 -m pip install yolov4
```

## Changelog

Ref: CHANGELOG


CHANGELOG
tensorflow-yolov4 (2.1.0) unstable; urgency=medium

  * tflite: add edgetpu_lib argument to load_tflite()

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 27 Jan 2021 11:09:02 +0900

tensorflow-yolov4 (2.0.3) unstable; urgency=medium

  * common: allow media_path to be video stream id
  * tflite: fix problem of not finding load_delegate()

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 07 Jan 2021 21:34:35 +0900

tensorflow-yolov4 (2.0.2) ustable; urgency=medium

  * tf: set converter.experimental_new_converter to False
  * tf: set converter.inference_in/output_type to tf.uint8
  * tflite: fix problem of not finding load_delegate()

 -- Hyeonki Hong <hhk7734@gmail.com>  Sun, 20 Dec 2020 13:06:08 +0900

tensorflow-yolov4 (2.0.1) unstable; urgency=medium

  * tf: set memory growth to True when using GPU
  * model: head: correct line ending
  * tf: add **kwargs parameter to compile() and fit()
  * tf: fix strange ground truth values issue

 -- Hyeonki Hong <hhk7734@gmail.com>  Sat, 05 Dec 2020 18:26:08 +0900

tensorflow-yolov4 (2.0.0) unstable; urgency=medium

  * tf:dataset: add data format check
  * common: base_class: change input_size from integer to (width, height)
  * common: base_class: set default arguments for candidates_to_pred_bboxes
  * tf: dataset: modify variable initialization according to input_size type change
  * common: predict: reflect input_size type change
  * common: media: reflect input_size type change to resize_image
  * tf:dataset: reflect input_size type change to data augmentation
  * common: base_class: reflect predict.fit_pred_bboxes_to_original change
  * common: predict: fix problem that occurred because image size change was not considered
  * tf: reflect input_size type change
  * model: head: reflect input_size type change
  * tf: train: reflect input_size type change
  * common: base_class: change strides to immutable constant
  * common: base_class: add cap.release() to release resources
  * tflite: reflect input_size type change
  * tf: optimize predict() using @tf.function
  * common: predict: fix problem caused by incorrect iou calculation

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 27 Oct 2020 00:11:59 +0900

tensorflow-yolov4 (1.2.1) unstable; urgency=medium

  * tf: set num_sample to the number of dataset, if not specified
  * tf: modify to bring images in order on post-training
  * common: base_class: add iou/score_threshold argument
  * tf: add images_optional argument in save_dataset_for_mAP()
  * common: base_class: modify to call cv2.namedWindow once
  * common: media: modify to create colors once
  * tf: dataset: update docs

 -- Hyeonki Hong <hhk7734@gmail.com>  Fri, 04 Sep 2020 20:21:56 +0900

tensorflow-yolov4 (1.2.0) unstable; urgency=medium

  * model: remove tpu argument
  * tf: reflect model changes
  * tflite: remove tpu_hair
  * tf, tflite: move inference into BaseClass
  * model: yolov4: remove unused import
  * pylint: update .pylintrc
  * common: base_class: modify inference() to be able to use cam
  * common: base_class: add fps display to screen

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 11 Aug 2020 21:43:29 +0900

tensorflow-yolov4 (1.1.0) unstable; urgency=medium

  * model: add kernel_regularizer argument
  * tf: remove subdivision
  * tf: dataset: fix issue where coordinates become 0 due to repeated division
  * tf: dataset: modify to shuffle when augmentation
  * tf: train: modify order of declaration
  * tf: train: give xiou_loss to more weight
  * tf: train: use keras.losses.BinaryCrossentropy to avoid nan problem
  * tf: train: add verbose argument
  * tf: dataset: change converted_coco format
  * tf: add image_path_prefix argument to load_dataset()
  * tf: add loss_verbose argument to compile()
  * tf: reflect dataset format change to save_dataset_for_mAP()

 -- Hyeonki Hong <hhk7734@gmail.com>  Mon, 10 Aug 2020 19:09:15 +0900

tensorflow-yolov4 (1.0.0) unstable; urgency=medium

  * tf: dataset: add `label smooting` argument
  * tf: add label_smooting argument to load_dataset()
  * tf: train: set truth_conf threshold to 0.5 because of label smoothing

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 05 Aug 2020 13:05:02 +0900

tensorflow-yolov4 (0.24.0) unstable; urgency=medium

  * tf: set default num of sample for mAP to 1000
  * tf: fix error handling when image is empty

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 04 Aug 2020 12:40:25 +0900

tensorflow-yolov4 (0.23.0) unstable; urgency=medium

  * tf: dataset: add error handling when bboxes are empty
  * common: media: replace space of class name to underbar
  * tf: implement save_dataset_for_mAP()
  * tf: dataset: add error handling when image is empty

 -- Hyeonki Hong <hhk7734@gmail.com>  Mon, 03 Aug 2020 20:49:51 +0900

tensorflow-yolov4 (0.22.0) unstable; urgency=medium

  * tf: add TFLITE_BUILTINS, SELECT_TF_OPS to default supported_ops
  * tf: dataset: modify to check if image exists when creating a dataset
  * tf: dataset: remove preprocess_dataset
  * tf: dataset: add cut_out
  * tf: dataset: add _next_random_augmentation_data()
  * tf: dataset: add mix_up
  * tf: dataset: implement mosaic

 -- Hyeonki Hong <hhk7734@gmail.com>  Sun, 02 Aug 2020 17:22:43 +0900

tensorflow-yolov4 (0.21.0) unstable; urgency=medium

  * tf: train: set epsilon used for division to 1e-8
  * tf: train: add giou and iou to iou_type
  * tf: train: fix loss function of bbox_probabilities

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 30 Jul 2020 05:18:12 +0900

tensorflow-yolov4 (0.20.0) unstable; urgency=medium

  * tf: train: modify epsilon to 1e-9
  * tf: train: remove weight for conf_noobj_loss
  * tf: add arguments to fit()
  * tf: dataset: fix problem of not finding images
  * tf: add SaveWeightsCallback

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 29 Jul 2020 05:06:25 +0900

tensorflow-yolov4 (0.19.0) unstable; urgency=medium

  * tf: weights: modify 'set' to 'load'
  * tf: weights: implement *_save_weights
  * tf: add save_weights()


core/dataset.py
#! /usr/bin/env python
# coding=utf-8
# ================================================================
#   Copyright (C) 2019 * Ltd. All rights reserved.
#
#   Editor      : VIM
#   File name   : dataset.py
#   Author      : YunYang1994
#   Created date: 2019-03-15 18:05:03
#   Description :
#
# ================================================================

import os
import cv2
import random
import numpy as np
import tensorflow as tf
import core.utils as utils
from core.config import cfg


class Dataset(object):
    """implement Dataset here"""

    def __init__(self, dataset_type):
        self.annot_path = (
            cfg.TRAIN.ANNOT_PATH
            if dataset_type == "train"
            else cfg.TEST.ANNOT_PATH
        )
        self.input_sizes = (
            cfg.TRAIN.INPUT_SIZE
            if dataset_type == "train"
            else cfg.TEST.INPUT_SIZE
        )
        self.batch_size = (
            cfg.TRAIN.BATCH_SIZE
            if dataset_type == "train"
            else cfg.TEST.BATCH_SIZE
        )
        self.data_aug = (
            cfg.TRAIN.DATA_AUG if dataset_type == "train" else cfg.TEST.DATA_AUG
        )

        self.train_input_sizes = cfg.TRAIN.INPUT_SIZE
        self.strides = np.array(cfg.YOLO.STRIDES)
        self.classes = utils.read_class_names(cfg.YOLO.CLASSES)
        self.num_classes = len(self.classes)
        self.anchors = np.array(utils.get_anchors(cfg.YOLO.ANCHORS))
        self.anchor_per_scale = cfg.YOLO.ANCHOR_PER_SCALE
        self.max_bbox_per_scale = 150

        self.annotations = self.load_annotations(dataset_type)
        self.num_samples = len(self.annotations)
        self.num_batchs = int(np.ceil(self.num_samples / self.batch_size))
        self.batch_count = 0

    def load_annotations(self, dataset_type):
        with open(self.annot_path, "r") as f:
            txt = f.readlines()
            annotations = [
                line.strip()
                for line in txt
                if len(line.strip().split()[1:]) != 0
            ]
        np.random.shuffle(annotations)
        return annotations

    def __iter__(self):
        return self

    def __next__(self):

        with tf.device("/cpu:0"):
            # self.train_input_size = random.choice(self.train_input_sizes)
            self.train_input_size = cfg.TRAIN.INPUT_SIZE
            self.train_output_sizes = self.train_input_size // self.strides

            batch_image = np.zeros(
                (
                    self.batch_size,
                    self.train_input_size,
                    self.train_input_size,
                    3,
                ),
                dtype=np.float32,
            )

            batch_label_sbbox = np.zeros(
                (
                    self.batch_size,
                    self.train_output_sizes[0],
                    self.train_output_sizes[0],
                    self.anchor_per_scale,
                    5 + self.num_classes,
                ),
                dtype=np.float32,
            )
            batch_label_mbbox = np.zeros(
                (
                    self.batch_size,
                    self.train_output_sizes[1],
                    self.train_output_sizes[1],
                    self.anchor_per_scale,
                    5 + self.num_classes,
                ),
                dtype=np.float32,
            )
            batch_label_lbbox = np.zeros(
                (
                    self.batch_size,
                    self.train_output_sizes[2],
                    self.train_output_sizes[2],
                    self.anchor_per_scale,
                    5 + self.num_classes,
                ),
                dtype=np.float32,
            )

            batch_sbboxes = np.zeros(
                (self.batch_size, self.max_bbox_per_scale, 4), dtype=np.float32
            )
            batch_mbboxes = np.zeros(
                (self.batch_size, self.max_bbox_per_scale, 4), dtype=np.float32
            )
            batch_lbboxes = np.zeros(
                (self.batch_size, self.max_bbox_per_scale, 4), dtype=np.float32
            )

            num = 0
            if self.batch_count < self.num_batchs:
                while num < self.batch_size:
                    index = self.batch_count * self.batch_size + num
                    if index >= self.num_samples:
                        index -= self.num_samples
                    annotation = self.annotations[index]
                    image, bboxes = self.parse_annotation(annotation)
                    (
                        label_sbbox,
                        label_mbbox,
                        label_lbbox,
                        sbboxes,
                        mbboxes,
                        lbboxes,
                    ) = self.preprocess_true_boxes(bboxes)

                    batch_image[num, :, :, :] = image
                    batch_label_sbbox[num, :, :, :, :] = label_sbbox
                    batch_label_mbbox[num, :, :, :, :] = label_mbbox
                    batch_label_lbbox[num, :, :, :, :] = label_lbbox
                    batch_sbboxes[num, :, :] = sbboxes
                    batch_mbboxes[num, :, :] = mbboxes
                    batch_lbboxes[num, :, :] = lbboxes
                    num += 1
                self.batch_count += 1
                batch_smaller_target = batch_label_sbbox, batch_sbboxes
                batch_medium_target = batch_label_mbbox, batch_mbboxes
                batch_larger_target = batch_label_lbbox, batch_lbboxes

                return (
                    batch_image,
                    (
                        batch_smaller_target,
                        batch_medium_target,
                        batch_larger_target,
                    ),
                )
            else:
                self.batch_count = 0
                np.random.shuffle(self.annotations)
                raise StopIteration

    def random_horizontal_flip(self, image, bboxes):

        if random.random() < 0.5:
            _, w, _ = image.shape
            image = image[:, ::-1, :]
            bboxes[:, [0, 2]] = w - bboxes[:, [2, 0]]

        return image, bboxes

    def random_crop(self, image, bboxes):

        if random.random() < 0.5:
            h, w, _ = image.shape
            max_bbox = np.concatenate(
                [
                    np.min(bboxes[:, 0:2], axis=0),
                    np.max(bboxes[:, 2:4], axis=0),
                ],
                axis=-1,
            )

            max_l_trans = max_bbox[0]
            max_u_trans = max_bbox[1]
            max_r_trans = w - max_bbox[2]
            max_d_trans = h - max_bbox[3]

            crop_xmin = max(
                0, int(max_bbox[0] - random.uniform(0, max_l_trans))
            )
            crop_ymin = max(
                0, int(max_bbox[1] - random.uniform(0, max_u_trans))
            )
            crop_xmax = max(
                w, int(max_bbox[2] + random.uniform(0, max_r_trans))
            )
            crop_ymax = max(
                h, int(max_bbox[3] + random.uniform(0, max_d_trans))
            )

            image = image[crop_ymin:crop_ymax, crop_xmin:crop_xmax]

            bboxes[:, [0, 2]] = bboxes[:, [0, 2]] - crop_xmin
            bboxes[:, [1, 3]] = bboxes[:, [1, 3]] - crop_ymin

        return image, bboxes

    def random_translate(self, image, bboxes):

        if random.random() < 0.5:
            h, w, _ = image.shape
            max_bbox = np.concatenate(
                [
                    np.min(bboxes[:, 0:2], axis=0),
                    np.max(bboxes[:, 2:4], axis=0),
                ],
                axis=-1,
            )

            max_l_trans = max_bbox[0]
            max_u_trans = max_bbox[1]
            max_r_trans = w - max_bbox[2]
            max_d_trans = h - max_bbox[3]

            tx = random.uniform(-(max_l_trans - 1), (max_r_trans - 1))
            ty = random.uniform(-(max_u_trans - 1), (max_d_trans - 1))

            M = np.array([[1, 0, tx], [0, 1, ty]])
            image = cv2.warpAffine(image, M, (w, h))

            bboxes[:, [0, 2]] = bboxes[:, [0, 2]] + tx
            bboxes[:, [1, 3]] = bboxes[:, [1, 3]] + ty

        return image, bboxes

    def parse_annotation(self, annotation):

        line = annotation.split()
        image_path = line[0]
        if not os.path.exists(image_path):
            raise KeyError("%s does not exist ... " % image_path)
        image = cv2.imread(image_path)
        bboxes = np.array([list(map(int, box.split(","))) for box in line[1:]])

        if self.data_aug:
            image, bboxes = self.random_horizontal_flip(
                np.copy(image), np.copy(bboxes)
            )
            image, bboxes = self.random_crop(np.copy(image), np.copy(bboxes))
            image, bboxes = self.random_translate(
                np.copy(image), np.copy(bboxes)
            )

        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        image, bboxes = utils.image_preporcess(
            np.copy(image),
            [self.train_input_size, self.train_input_size],
            np.copy(bboxes),
        )
        return image, bboxes

    def bbox_iou(self, boxes1, boxes2):

        boxes1 = np.array(boxes1)
        boxes2 = np.array(boxes2)

        boxes1_area = boxes1[..., 2] * boxes1[..., 3]
        boxes2_area = boxes2[..., 2] * boxes2[..., 3]

        boxes1 = np.concatenate(
            [
                boxes1[..., :2] - boxes1[..., 2:] * 0.5,
                boxes1[..., :2] + boxes1[..., 2:] * 0.5,
            ],
            axis=-1,
        )
        boxes2 = np.concatenate(
            [
                boxes2[..., :2] - boxes2[..., 2:] * 0.5,
                boxes2[..., :2] + boxes2[..., 2:] * 0.5,
            ],
            axis=-1,
        )

        left_up = np.maximum(boxes1[..., :2], boxes2[..., :2])
        right_down = np.minimum(boxes1[..., 2:], boxes2[..., 2:])

        inter_section = np.maximum(right_down - left_up, 0.0)
        inter_area = inter_section[..., 0] * inter_section[..., 1]
        union_area = boxes1_area + boxes2_area - inter_area

        return inter_area / union_area

    def preprocess_true_boxes(self, bboxes):

        label = [
            np.zeros(
                (
                    self.train_output_sizes[i],
                    self.train_output_sizes[i],
                    self.anchor_per_scale,
                    5 + self.num_classes,
                )
            )
            for i in range(3)
        ]
        bboxes_xywh = [np.zeros((self.max_bbox_per_scale, 4)) for _ in range(3)]
        bbox_count = np.zeros((3,))

        for bbox in bboxes:
            bbox_coor = bbox[:4]
            bbox_class_ind = bbox[4]

            onehot = np.zeros(self.num_classes, dtype=np.float)
            onehot[bbox_class_ind] = 1.0
            uniform_distribution = np.full(
                self.num_classes, 1.0 / self.num_classes
            )
            deta = 0.01
            smooth_onehot = onehot * (1 - deta) + deta * uniform_distribution

            bbox_xywh = np.concatenate(
                [
                    (bbox_coor[2:] + bbox_coor[:2]) * 0.5,
                    bbox_coor[2:] - bbox_coor[:2],
                ],
                axis=-1,
            )
            bbox_xywh_scaled = (
                1.0 * bbox_xywh[np.newaxis, :] / self.strides[:, np.newaxis]
            )

            iou = []
            exist_positive = False
            for i in range(3):
                anchors_xywh = np.zeros((self.anchor_per_scale, 4))
                anchors_xywh[:, 0:2] = (
                    np.floor(bbox_xywh_scaled[i, 0:2]).astype(np.int32) + 0.5
                )
                anchors_xywh[:, 2:4] = self.anchors[i]

                iou_scale = self.bbox_iou(
                    bbox_xywh_scaled[i][np.newaxis, :], anchors_xywh
                )
                iou.append(iou_scale)
                iou_mask = iou_scale > 0.3

                if np.any(iou_mask):
                    xind, yind = np.floor(bbox_xywh_scaled[i, 0:2]).astype(
                        np.int32
                    )

                    label[i][yind, xind, iou_mask, :] = 0
                    label[i][yind, xind, iou_mask, 0:4] = bbox_xywh
                    label[i][yind, xind, iou_mask, 4:5] = 1.0
                    label[i][yind, xind, iou_mask, 5:] = smooth_onehot

                    bbox_ind = int(bbox_count[i] % self.max_bbox_per_scale)
                    bboxes_xywh[i][bbox_ind, :4] = bbox_xywh
                    bbox_count[i] += 1

                    exist_positive = True

            if not exist_positive:
                best_anchor_ind = np.argmax(np.array(iou).reshape(-1), axis=-1)
                best_detect = int(best_anchor_ind / self.anchor_per_scale)
                best_anchor = int(best_anchor_ind % self.anchor_per_scale)
                xind, yind = np.floor(
                    bbox_xywh_scaled[best_detect, 0:2]
                ).astype(np.int32)

                label[best_detect][yind, xind, best_anchor, :] = 0
                label[best_detect][yind, xind, best_anchor, 0:4] = bbox_xywh
                label[best_detect][yind, xind, best_anchor, 4:5] = 1.0
                label[best_detect][yind, xind, best_anchor, 5:] = smooth_onehot

                bbox_ind = int(
                    bbox_count[best_detect] % self.max_bbox_per_scale
                )
                bboxes_xywh[best_detect][bbox_ind, :4] = bbox_xywh
                bbox_count[best_detect] += 1
        label_sbbox, label_mbbox, label_lbbox = label
        sbboxes, mbboxes, lbboxes = bboxes_xywh
        return label_sbbox, label_mbbox, label_lbbox, sbboxes, mbboxes, lbboxes

    def __len__(self):
        return self.num_batchs
